src/edsger/core.cljs

```diff
@@ -50,32 +56,85 @@
+         "<button type=\"button\" class=\"close\" data-dismiss=\"alert\" aria-label=\"Close\">"
+           "<span aria-hidden=\"true\">&times;</span>"
+         "</button>"
+         "Error in input"
+       "</div>"))
+
+(defn- merge-val
+  "Takes a map looking `{:curr-id # :locations []}` and a number.
+   Returns a new map looking `{:curr-id (inc #) :locations []}`.
+   If val is nil, (inc #) is added to the `:locations` vector"
+  [curr-map val]
+  (let [curr-map (update curr-map :curr-id inc)
+        curr-id (:curr-id curr-map)]
+    (if (nil? val) (update-in curr-map [:locations] #(conj % curr-id)) curr-map)))
+
+(defn- show-exp-parse-err
+  "Takes a vector of indices and attach error message elements
+   to corresponding exp input box"
+  [err-id-vec]
+  (let [exp-boxes (iArrayLike-to-cljs-list (gdom/getElementsByClass "exp-box"))]
+    (dorun
+     (map
+      (fn [id]
+        ;; adding err msg where they fail
+        (gdom/appendChild (nth exp-boxes id) (str-to-elem parse-err-str)))
+      err-id-vec))))
+
+(defn- show-rule-parse-err
+  "Takes a vector of indices and attach error message elements
+   to corresponding rule input box"
+  [err-id-vec]
+  (let [rule-cols (map #(gdom/getParentElement %)
+                       (iArrayLike-to-cljs-list (gdom/getElementsByClass "rule")))]
+    (dorun
+     (map
+      (fn [id]
+        ;; adding err msg where they fail
+        (gdom/insertSiblingAfter (str-to-elem parse-err-str) (nth rule-cols id)))
+      err-id-vec))))
+
 (defn validate-handler
   "Performs the validation based on the values typed by users"
   [evt]
-  (let [ex-elems (gdom/getElementsByClass "ex") ;; `iArrayLike` type
-        rule-elems (gdom/getElementsByClass "rule")
-        exps (map #(parsing/parse (.-value %)) (iArrayLike-to-cljs-list ex-elems))
-        rules (map #(parsing/rulify (parsing/parse (.-value %))) (iArrayLike-to-cljs-list rule-elems))
+  (let [exp-str-li (map #(.-value %) (iArrayLike-to-cljs-list (gdom/getElementsByClass "ex")))
+        rule-str-li (map #(.-value %) (iArrayLike-to-cljs-list (gdom/getElementsByClass "rule")))
+        non-empty-input (every? #(not= "" %) (concat exp-str-li rule-str-li))
+        exps (map #(parsing/parse %) exp-str-li)
+        vanilla-rules (map #(parsing/parse %) rule-str-li)
+        rules (map #(parsing/rulify %) vanilla-rules)
+        ;; vector of indices where parsing err occurred (e.g., [0, 3])
+        exp-parse-err (:locations (reduce merge-val {:curr-id -1 :locations []} exps))
+        rule-parse-err (:locations (reduce merge-val {:curr-id -1 :locations []} vanilla-rules))
         result-str (str (true? (uni/check-match-recursive (nth exps 0)
                                                           (nth exps 1)
                                                           (nth rules 0)
                                                           (nth rules 1))))]
-    (.alert js/window result-str)))
+    (when non-empty-input
+      (cond
```

Comment:
The way this `cond` is set up, users have to resolve all the parse errors in the start & end expressions before the site tells them that there is anything wrong with the rules.

Reply:
Fixed